proxy/proxy/client.go

```diff
@@ -292,33 +241,129 @@ func isExpired(cfg *tls.Config) bool {
+	return c.err == nil && c.cfg != nil
+}
+
+func needsRefresh(e cacheEntry, refreshCfgBuffer time.Duration) bool {
+	if e.done == nil { // no refresh started
+		return true
+	}
+	if !isValid(e) || time.Since(e.lastRefreshed) >= refreshCfgBuffer {
+		// if the entry is invalid or close enough to expiring check
+		// use the entry's done channel to determine if a refresh has started yet
+		select {
+		case <-e.done: // last refresh completed, so it's time for a new one
+			return true
+		default: // new refresh already started, so we can wait on that
+			return false
+		}
+	}
+	return false
+}
+
+func (c *Client) cachedCfg(instance string) (string, *tls.Config, string, error) {
 	c.cacheL.RLock()
-	ret, ok := c.cfgCache[instance]
+
+	throttle := c.RefreshCfgThrottle
+	if throttle == 0 {
+		throttle = DefaultRefreshCfgThrottle
+	}
+	refreshCfgBuffer := c.RefreshCfgBuffer
+	if refreshCfgBuffer == 0 {
+		refreshCfgBuffer = defaultRefreshCfgBuffer
+	}
+
+	e := c.cfgCache[instance]
 	c.cacheL.RUnlock()
+	if needsRefresh(e, refreshCfgBuffer) {
+		// Reenter the critical section with intent to make changes
+		c.cacheL.Lock()
+		if c.cfgCache == nil {
+			c.cfgCache = make(map[string]cacheEntry)
+		}
+		// the state may have changed between critical sections, so double check
+		e = c.cfgCache[instance]
+		if needsRefresh(e, refreshCfgBuffer) {
+			if time.Since(e.lastRefreshed) >= throttle {
+				// start a new refresh and update the cachedEntry to reflect that
+				e.done = c.startRefresh(instance, refreshCfgBuffer)
+				e.lastRefreshed = time.Now()
+				c.cfgCache[instance] = e
+			} else {
+				// TODO: Should this be a returned error?
+				logging.Errorf("Throttling refreshCfg(%s): it was only called %v ago", instance, time.Since(e.lastRefreshed))
+			}
+		}
+		c.cacheL.Unlock()
+	}
 
-	// Don't waste time returning an expired/invalid cert.
-	if !ok || ret.err != nil || isExpired(ret.cfg) {
-		return "", nil, ""
+	if !isValid(e) {
+		// if the previous result was invalid, wait for the next result to complete
+		// TODO: Ask for context?
+		<-e.done
+		c.cacheL.RLock()
+		// the state may have changed between critical sections, so double check
+		e = c.cfgCache[instance]
+		c.cacheL.RUnlock()
 	}
-	return ret.addr, ret.cfg, ret.version
+	return e.addr, e.cfg, e.version, e.err
 }
 
 // DialContext uses the configuration stored in the client to connect to an instance.
 // If this func returns a nil error the connection is correctly authenticated
 // to connect to the instance.
 func (c *Client) DialContext(ctx context.Context, instance string) (net.Conn, error) {
-	if addr, cfg, _ := c.cachedCfg(instance); cfg != nil {
-		ret, err := c.tryConnect(ctx, addr, cfg)
-		if err == nil {
-			return ret, err
-		}
-	}
-
-	addr, cfg, _, err := c.refreshCfg(instance)
+	addr, cfg, _, err := c.cachedCfg(instance)
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO: attempt an early refresh if an connect fails?
```

Comment:
I think we can hold off on this for now until we better understand how this code performs generally.